server/src/main/java/org/elasticsearch/search/query/QueryPhaseCollectorManager.java

```diff
@@ -0,0 +1,714 @@
+        boolean profile,
+        CollapseContext collapseContext,
+        @Nullable SortAndFormats sortAndFormats,
+        int numHits,
+        boolean trackMaxScore,
+        @Nullable FieldDoc after
+    ) {
+        assert numHits > 0;
+        assert collapseContext != null;
+        Sort sort = sortAndFormats == null ? Sort.RELEVANCE : sortAndFormats.sort;
+        final SinglePassGroupingCollector<?> topDocsCollector = collapseContext.createTopDocs(sort, numHits, after);
+        MaxScoreCollector maxScoreCollector = trackMaxScore ? new MaxScoreCollector() : null;
+        return new QueryPhaseCollectorManager(postFilterWeight, terminateAfterChecker, aggsCollectorManager, minScore, profile) {
+            boolean newCollectorCalled = false;
+
+            @Override
+            protected Collector newTopDocsCollector() {
+                assert newCollectorCalled == false : "Field collapsing does not support concurrent execution";
+                newCollectorCalled = true;
+                return MultiCollector.wrap(topDocsCollector, maxScoreCollector);
+            }
+
+            @Override
+            protected TopDocsAndMaxScore reduceTopDocsCollectors(Collection<Collector> collectors) throws IOException {
+                assert collectors.size() == 1 : "Field collapsing does not support concurrent execution";
+                TopFieldGroups topDocs = topDocsCollector.getTopGroups(0);
+                float maxScore = getMaxScore(topDocs, sortAndFormats, trackMaxScore, Collections.singletonList(maxScoreCollector));
+                return new TopDocsAndMaxScore(topDocs, maxScore);
+            }
+
+            @Override
+            protected DocValueFormat[] getSortValueFormats() {
+                return sortAndFormats == null ? new DocValueFormat[] { DocValueFormat.RAW } : sortAndFormats.formats;
+            }
+        };
+    }
+
+    private static float getMaxScore(
+        TopDocs topDocs,
+        SortAndFormats sortAndFormats,
+        boolean trackMaxScore,
+        Collection<MaxScoreCollector> maxScoreCollectors
+    ) {
+        if (sortAndFormats == null) {
+            return topDocs.scoreDocs.length == 0 ? Float.NaN : topDocs.scoreDocs[0].score;
+        }
+        if (trackMaxScore) {
+            return maxScoreCollectors.stream().map(MaxScoreCollector::getMaxScore).reduce(Float.NEGATIVE_INFINITY, (f1, f2) -> {
+                if (Float.isNaN(f1)) {
+                    return f2;
+                }
+                if (Float.isNaN(f2)) {
+                    return f1;
+                }
+                return Math.max(f1, f2);
+            });
+        }
+        return Float.NaN;
+    }
+
+    /**
+     * Returns query total hit count if the <code>query</code> is a {@link MatchAllDocsQuery}
+     * or a {@link TermQuery} and the <code>reader</code> has no deletions,
+     * -1 otherwise.
+     */
+    static int shortcutTotalHitCount(IndexReader reader, Query query) throws IOException {
```

Comment:
this is nothing new, it was copied with no changes form TopDocsCollectorManagerFactory. Nothing to look at.